week-8/handout.md

```diff
@@ -0,0 +1,190 @@
+    try (BufferedReader br = new BufferedReader(new FileReader(path))) {
+        return br.readLine();
+    }
+}
+```
+
+#### Throw
+A `throw` kulcsszóval tudunk kivétel példányokat eldobni.
+```Java
+public void doSomething() {
+    throw new IllegalArgumentException();
+}
+```
+
+### Saját kivétel osztály létrehozása
+Könnyedén hozhatunk létre saját (custom) kivételeket. Egyszerűen csak származtatnunk kell az Exception osztályt (ha
+checked exception-t szeretnénk) vagy a RuntimeException osztályt (ha unchecked exception-t szeretnénk).
+
+## Annotációk
+Az annotációk segítségével különböző nyelvi elemeket (osztály, mező, metódus, paraméter, stb) tudunk megjelölni, vagy más
+szóval megannotálni. Ezeket az annotációk szólhatnak a fejlesztőknek, de akár a fordító vagy esetleges posztprocesszorok
+is feldolgozhatják.
+
+Ilyen annotációra egy példa a már korábban is látott `@Override`, ami azt jelzi a fordítónak, hogy az adott metódust
+(amire rátettük) azzal a céllal hoztuk létre, hogy az ősosztály egy metódusát felüldefiniáljuk.
+```java
+public class Car {
+    @Override
+    public String toString() {
+        return "Car";
+    }
+}
+```
+
+### Saját Annotáció létrehozása
+Saját annotációt az alábi módon hozhatunk létre:
+```java
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.TYPE, ElementType.METHOD})
+public @interface MyAnnotation {
+	String value() default "";
+	int number() default 0;
+}
+```
+Az annotációknak lehetnek attribútumai. Az előző példában a `@MyAnnotation` annotációnak két attribútuma van, a `value`
+és a `number`. EZeknek az attribútumoknak a típusa az alábbiak lehetnek: primitív típus (int, long, stb), String, Class,
+Enum, Annotáció vagy ilyen típusú tömb.
+
+### @Retention
+A `@Retention` egy speciális anniotáció, amit annotációk definiálására használunk. Azt mondja meg, hogy az adott annotáció
+meddig lesz elérhető, meddig lesz megtartva. Egy attribútuma van ami egy RetentionPolicy Enum, melynek három lehetséges
+értéke van:
+* SOURCE: az adott annotáció a fordítás során elveszik, a lefordított `.class` állományban már nem lesz benne
+* CLASS: a lefordított `.class` állományba még belekerül, de futásidőben a JVM által már nem lesz látható
+* RUNTIME: a lefordított `.class` állományba bele fog kerül, és futásidőben a JVM által lekérdezhető lesz
+
+### @Target
+A `@Target` egy másik speciális anniotáció, amit szintén annotációk definiálására használunk. Segítségével megadhatjuk,
+hogy milyen elemekre lehet az adott annotációt rátenni. Pl.: TYPE, FIELD, METHOD, PARAMETER stb. Az előző példában
+szereplő `@MyAnnotation` annotációt tehát osztályokra és metódusokra lehet rátenni.
+
+### Annotációk használata
+Az előző példában létrehozott `@MyAnnotation` annotációt az alábbi módokon használhatjuk és paranméterezhetjük:
+```java
+@MyAnnotation("asd")
+public class MyClass {
+	@MyAnnotation(number = 2)
+	@Override
+	public boolean equals(Object obj) {
+		return super.equals(obj);
+	}
+}
+``` 
+
+## Reflexió
```

Comment:
Shall we have 'Reflection' here as well or in a separate document?

Reply:
Discuss i person, that Reflection topic can be part of this handout too